File: Turquoise/UI/VC/TQGroupSelectionViewController.m
#import "TQGroupSelectionViewController.h"

#import "TQGroupSelectionTableViewCell.h"
#import "TQNNTPManager.h"
#import "TQUserInfoManager.h"

@class TQSearchBar;

@implementation TQGroupSelectionViewController {
  IBOutlet UITableView *_tableView;
  IBOutlet TQSearchBar *_searchBar;

  TQNNTPManager *_nntpManager;
  NSArray<TQNNTPGroup *> *_filteredGroups;
}

- (IBAction)doneButtonDidTap:(id)sender {
  [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad {
  [super viewDidLoad];

  _searchBar.tq_textField.font = [UIFont fontWithName:@"dungeon" size:12];
  _searchBar.autocapitalizationType = UITextAutocapitalizationTypeNone;

  _nntpManager = [TQNNTPManager sharedInstance];
  [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(newsGroupsDidUpdate:)
                                               name:kNNTPGroupListDidUpdateNotification
                                             object:nil];
}

- (void)newsGroupsDidUpdate:(NSNotification *)notification {
  [_tableView reloadData];
}

- (NSArray<TQNNTPGroup *> *)groups {
  return _filteredGroups ?: _nntpManager.allGroups;
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  return [self groups].count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  TQGroupSelectionTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"GroupCell"
                                                                       forIndexPath:indexPath];
  cell.group = [self groups][indexPath.row];
  cell.contentView.backgroundColor = (indexPath.row % 2 == 0)
      ? [TQGroupSelectionTableViewCell evenColor]
      : [TQGroupSelectionTableViewCell oddColor];
  return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  TQNNTPGroup *group = [self groups][indexPath.row];
  TQUserInfoManager *userInfoManager = [TQUserInfoManager sharedInstance];
  if ([userInfoManager isSubscribedToGroup:group]) {
    [userInfoManager unsubscribeFromGroup:group];
  } else {
    [userInfoManager subscribeToGroup:group];
  }

  [tableView deselectRowAtIndexPath:indexPath animated:NO];
}

#pragma mark - UISearchBarDelegate

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
  if (searchText.length == 0) {
    _filteredGroups = nil;
  } else {
    NSPredicate *predicate =
        [NSPredicate predicateWithBlock:^BOOL(TQNNTPGroup *_Nullable evaluatedGroup, NSDictionary *_Nullable bindings) {
          return [evaluatedGroup.groupId localizedCaseInsensitiveContainsString:searchText];
        }];
    _filteredGroups = [_nntpManager.allGroups filteredArrayUsingPredicate:predicate];
  }

  [_tableView reloadData];
}



/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: Turquoise/UI/VC/TQGroupSelectionViewController.swift
import Foundation
import UIKit

class TQGroupSelectionViewController : UIViewController, UISearchBarDelegate, UITableViewDataSource, UITableViewDelegate {
  @IBOutlet var tableView: UITableView!
  @IBOutlet var searchBar: TQSearchBar!
  var filteredGroups: [TQNNTPGroup]?

  @IBAction func doneButtonDidTap(_ sender: Any?) {
    self.dismiss(animated: true, completion: nil)
  }

  override func viewDidLoad() {
    super.viewDidLoad()

    self.searchBar.tq_textField?.font = UIFont(name: "dungeon", size: 12)
    self.searchBar.autocapitalizationType = .none

    NotificationCenter.default.addObserver(self,
                                           selector: #selector(newsGroupsDidUpdate(_:)),
                                           name: TQNNTPManager.NNTPGroupListDidUpdateNotification,
                                           object: nil)
  }


  func newsGroupsDidUpdate(_ notification: Notification) {
    self.tableView.reloadData()
  }

  func groups() -> [TQNNTPGroup] {
    return self.filteredGroups ?? TQNNTPManager.sharedInstance.allGroups
  }

  // MARK: - UITableViewDataSource

  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return self.groups().count
  }

  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "GroupCell", for: indexPath) as! TQGroupSelectionTableViewCell
    cell.group = self.groups()[indexPath.row]
    cell.contentView.backgroundColor = (indexPath.row % 2 == 0)
      ? TQGroupSelectionTableViewCell.evenColor()
      : TQGroupSelectionTableViewCell.oddColor()
    return cell
  }

  // MARK: - UITableViewDelegate

  func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    let group = self.groups()[indexPath.row]
    let userInfoManager = TQUserInfoManager.sharedInstance
    if userInfoManager.isSubscribedTo(group: group) {
      userInfoManager.unsubscribeFrom(group: group)
    } else {
      userInfoManager.subscribeTo(group: group)
    }

    tableView.deselectRow(at: indexPath, animated: false)
  }

  // MARK: - UISearchBarDelegate

  func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
    if searchText.isEmpty {
      self.filteredGroups = nil
    } else {
      self.filteredGroups = TQNNTPManager.sharedInstance.allGroups.filter{ (group) -> Bool in
        group.groupId.localizedCaseInsensitiveContains(searchText)
      }
    }

    tableView.reloadData()
  }
}
